examples/experimental/rusty_pipes/rusty_pipes/rust/src/main.rs

```diff
@@ -0,0 +1,202 @@
+        writeln!(
+            &mut self.file,
+            "{}",
+            serde_json::to_string(&message).unwrap()
+        );
+    }
+}
+
+#[derive(Debug)]
+struct PipesContext {
+    data: PipesContextData,
+    writer: PipesFileMessageWriter, // should be trait & support stdio
+}
+
+impl PipesContext {
+    fn report_asset_check(
+        &mut self,
+        check_name: &str,
+        passed: bool,
+        asset_key: &str,
+        metadata: serde_json::Value,
+    ) {
+        let params: HashMap<String, serde_json::Value> = HashMap::from([
+            ("asset_key".to_string(), json!(asset_key)),
+            ("check_name".to_string(), json!(check_name)),
+            ("passed".to_string(), json!(passed)),
+            ("severity".to_string(), json!("ERROR")), // hardcode for now
+            ("metadata".to_string(), metadata),
+        ]);
+
+        let msg = PipesMessage {
+            __dagster_pipes_version: "0.1".to_string(),
+            method: "report_asset_check".to_string(),
+            params: Some(params),
+        };
+        self.writer.write_message(msg);
+    }
+}
+
+fn load_pipes_param<T>(param: &str) -> T
+where
+    T: de::DeserializeOwned,
+{
+    let zlib_compressed_slice = base64::decode(param).unwrap();
+    let mut decoder = ZlibDecoder::new(&zlib_compressed_slice[..]);
+    let mut json_str = String::new();
+    decoder.read_to_string(&mut json_str).unwrap();
+    let value: T = serde_json::from_str(&json_str).unwrap();
+    return value;
+}
+
+fn load_pipes_context_data(param: &str) -> PipesContextData {
+    let context_payload: PipesContextPayload = load_pipes_param(&param);
+
+    if let Some(context_data) = context_payload.data {
+        return context_data;
+    } else if let Some(context_path) = context_payload.path {
+        let file = File::open(context_path).unwrap();
+        let reader = BufReader::new(file);
+        let data: PipesContextData = serde_json::from_reader(reader).unwrap();
+        return data;
+    } else {
+        panic!("unable to load dagster pipes context data")
+    }
+}
+
+fn load_pipes_message_writer(param: &str) -> PipesFileMessageWriter {
+    let messages_config: PipesMessagesConfig = load_pipes_param(&param);
+    if let Some(path) = messages_config.path {
+        return PipesFileMessageWriter::new(&path);
+    }
+    panic!("Unable to load pipes message writer. Only file based is currently supported")
+}
+
+fn get_dagster_pipes_context() -> PipesContext {
+    let context_param = env::var("DAGSTER_PIPES_CONTEXT").unwrap();
+    let context_data = load_pipes_context_data(&context_param);
+    let messages_param = env::var("DAGSTER_PIPES_MESSAGES").unwrap();
+    let message_writer = load_pipes_message_writer(&messages_param);
+    return PipesContext {
+        data: context_data,
+        writer: message_writer,
+    };
+}
+
+fn main() {
+    let mut context = get_dagster_pipes_context();
+    let metadata = json!({"null_count": {"raw_value": 0, "type": "int"}});
```

Comment:
ended up coercing to the "normalized" metadata format at the callsite instead of trying to figure out how to resolve it in rust. 

Looking more closely at the code that errored maybe i did something else stupid.